src/naip_cnn/sampling.py

```diff
@@ -1,227 +1,89 @@
+    y = geom.transform(proj).coordinates().getNumber(1)
+
+    xmin = x.subtract(dims[0] // 2)
+    ymin = y.subtract(dims[1] // 2)
+    xmax = x.add(dims[0] // 2)
+    ymax = y.add(dims[1] // 2)
+
+    footprint = ee.Geometry.Rectangle(
+        [xmin, ymin, xmax, ymax], proj=proj, geodesic=False
     )
+    # TODO: add id to footprint
+    return ee.Feature(footprint, {"height": dims[0], "width": dims[1]})
 
 
-def extract_footprints_from_dataframe(
+def extract_values_at_footprint(
+    footprint: ee.Feature,
     *,
-    df: gpd.GeoDataFrame,
     img: ee.Image,
-    dims: tuple[int, int],
     proj: ee.Projection,
-    chunk_size: int = 5_000,
-) -> pd.DataFrame:
-    """Extract pixel values from an image over a dataframe of Point geometries.
-
-    Parameters
-    ----------
-    df : pd.DataFrame
-        A dataframe of Point geometries to sample values around.
-    img : ee.Image
-        The image to sample from.
-    dims : tuple[int, int]
-        The dimensions of the footprint to sample at each geometry, in the proj units.
-    proj : ee.Projection
-        The projection to sample in.
-    chunk_size : int
-        The number of geometries to sample at a time. This is used to avoid exceeding
-        Earth Engine limits. Because samples are downloaded in parallel, experimenting
-        with this value may improve performance.
-
-    Returns
-    -------
-    pd.DataFrame
-        A dataframe of pixel values from the image. Each row is a point from the
-        original dataframe, with one column added per image band. Pixel values are
-        stored as flat lists of values.
-    """
-    n_chunks = math.ceil(len(df) / chunk_size)
-
-    def _extract_footprints_from_chunk(chunk: pd.DataFrame) -> gpd.GeoDataFrame:
-        features = []
-        for _, f in chunk.iterrows():
-            footprint = _create_footprint_at_xy(
-                xy=(f.geometry.x, f.geometry.y), dims=dims, proj=proj
-            )
-            values = _extract_values_at_footprint(img=img, geom=footprint, proj=proj)
-            features.append(ee.Feature(None, values.set("id", f.id)))
-
-        features = ee.FeatureCollection(features)
-        info = features.getInfo()
-        return gpd.GeoDataFrame.from_features(info["features"])
-
-    chunks = (
-        df.iloc[chunk_size * chunk : chunk_size * (chunk + 1)]
-        for chunk in range(n_chunks)
-    )
-    gdfs = Parallel(n_jobs=-1, prefer="threads")(
-        delayed(_extract_footprints_from_chunk)(chunk) for chunk in chunks
+    scale: int = 1,
+    drop_if_null: bool = True,
+    **kwargs,
+) -> ee.Feature:
+    """Extract a footprint of pixel values from an image over a geometry."""
+    values = img.reduceRegion(
+        reducer=ee.Reducer.toList(),
+        geometry=footprint.geometry(),
+        scale=scale,
+        crs=proj,
+        **kwargs,
     )
 
-    return pd.concat(gdfs).drop(columns="geometry").reset_index(drop=True)
+    if drop_if_null:

```

Comment:
I *think* I understand how `drop_if_null` cooperates with the `ee.FeatureCollection.map` parameter `opt_dropNulls`, but probably good for me to talk it through.  If you pass `drop_if_null=True`, any footprint with masked pixels will get a value of None (null) and then `opt_dropNulls` will remove it from the feature collection.  But if you pass `drop_if_null=False`, these partially masked footprints will stay as features.  Will this cause issues when training?

Reply:
Your understanding is spot on. I'm pretty sure that the model would train just fine with nulls, but I'm not sure about the performance implications. Since we have no shortage of training data and nulls only tend to occur at the edges of the LiDAR data, I figured it's safer just to drop them

Reply:
I was actually thinking that more nulls might get caused by the LCMS mask, but perhaps that's not a huge amount considering the short time frame between NAIP and lidar acquisitions.  But, as you say, you're definitely not lacking for data.

Reply:
I wasn't thinking about that, but it is of course another source of null values! I don't have a great idea of how many footprints we're losing to that, but it would be good to know - I'll give that a quick look. Another thing to consider is that LCMS likely misses disturbances at the sub-30m pixel scale, so avoiding footprints that are immediately adjacent to masked pixels is probably a safer tactic, as long as we're not limited in our training data.

Reply:
Interesting, a bit higher loss rate than I would have guessed, but 19K *still* seems like a lot of training data.